### PR TITLE
FEEC with Dirichlet BCs.

### LINKED ISSUE #80: Add magic methods to the BlockMatrix

### PR DESCRIPTION
**Pull request by yguclu**
Apply homogeneous Dirichlet boundary conditions to V0 space of de Rham sequence. This PR also fixes #80.

**New features**
- Add option to 1D Maxwell test: Dirichlet BCs can be applied by penalizing the mass-matrix bilinear form, or strongly;
- Add 2D Maxwell test, where Dirichlet BCs are applied by penalizing the mass-matrix bilinear form;
- Add magic methods to `Matrix` and `Vector` base classes;
- Make Jacobi preconditioner work on block matrices, and for any number of dimensions;
- Change behavior of `BlockLinearOperator` for single row or column:
   - If no. of rows == 1, codomain does not have to be `ProductSpace`;
   - If no. of columns == 1 domain does not have to be `ProductSpace`;
   - Accordingly, dot product allows non-block vectors as input or output;
   - Further, `ProductSpace` constructed from one space returns input space;
   - This allows simplified usage of FEEC derivative matrices.

**Bug fixes**
- Fix bug in 2D push-forwards;
- Fix bug in constructors of `DiscreteLinearForm` and `DiscreteBilinearForm`;
- Fix bug in `apply_essential_bc_1d_StencilMatrix`;

**Improvements**
- Speed up Jacobi preconditioner;
- Reduce number of temporary vectors in CG and PCG solvers;
- Correct iteration counter in CG and PCG solvers.

**Other changes**
- Cleanup and document global projectors;
- Some cleanup in 2D push-forwards;
- Use SymPDE version 0.10.8.

### CODE REVIEW

File: psydac/api/tests/test_api_feec_2d.py [hunk 1]
```diff
@@ -0,0 +1,659 @@
...
+    # Solution
+    #--------------------------------------------------------------------------
+
+    # TODO: add option to convert to scipy sparse format
+
+    # Arguments for time stepping
+    if periodic:
+        args = (e, b, M1, M2, D1, D1_T)
+    else:
+        args = (e, b, M1 + M1_bc, M2, D1, D1_T)
```
Comment (campospinto) [inline psydac/api/tests/test_api_feec_2d.py:381]: the tangential trace mass matrix should probably be multiplied by a large penalization number

File: psydac/feec/global_projectors.py [hunk 2]
```diff
@@ -110,7 +110,7 @@ def __init__(self, Hcurl, n_quads=None):
```
Comment (campospinto) [inline psydac/feec/global_projectors.py:110]: the default None value for n_quads yields an error on L78 (same for the Hdiv projector)

File: psydac/feec/global_projectors.py [hunk 3]
```diff
@@ -275,11 +275,11 @@ def __init__(self, L2, quads=None):
         slices     = tuple(slice(p+1,-p-1) for p in L2.degree)
```
Comment (campospinto) [inline psydac/feec/global_projectors.py:275]: L2 projector has a 'quads' argument while Hcurl and Hdiv projs have 'n_quads'  arguments. Why ?

File: psydac/feec/global_projectors.py [hunk 4]
```diff
@@ -275,11 +275,11 @@ def __init__(self, L2, quads=None):
...
-            self.func = evaluate_dof_3form_1d
+            self.func = evaluate_dofs_1d_1form
         elif len(self.D) == 2:
-            self.func = evaluate_dof_3form_2d
+            self.func = evaluate_dofs_2d_2form
         elif len(self.D) == 3:
-            self.func = evaluate_dof_3form_3d
+            self.func = evaluate_dofs_3d_3form
         else:
             raise ValueError('H1 projector of dimension {} not available'.format(str(len(self.N))))
```
Comment (campospinto) [inline psydac/feec/global_projectors.py:284]: typo ? 'H1 projector' -> 'L2 projector'

File: psydac/feec/global_projectors.py [hunk 5]
```diff
@@ -275,11 +275,11 @@ def __init__(self, L2, quads=None):
...
```
Comment (campospinto) [inline psydac/feec/global_projectors.py:285]: on L290 it should not be 'space V1' as the index for an L2 projection is n = dimension
(same comment for "1-form" on L295)

File: psydac/feec/global_projectors.py [hunk 6]
```diff
@@ -313,23 +313,42 @@ def __call__(self, fun):
         return FemField(self.space, coeffs=coeffs)
 
 #==============================================================================
-def evaluate_dof_0form_1d(n1, points_1, F, f):
+# 1D DEGREES OF FREEDOM
```
Comment (campospinto) [inline psydac/feec/global_projectors.py:316]: comment should probably indicate whether `f` is a form in the logical or physical domain

File: psydac/feec/global_projectors.py [hunk 7]
```diff
@@ -29,11 +29,11 @@ def __init__(self, H1):
         self.args  = (*n_basis, *points, self.rhs._data[slices])
```
Comment (campospinto) [inline psydac/feec/global_projectors.py:30]: Projector docstrings should probably indicate whether they are to be applied to functions (forms) in the logical or physical domain

**Review by campospinto — COMMENTED** [review]
the 2D Maxwell solver in the test_api_feec looks very nice!  I left a few questions but mostly in the global_projectors module

File: psydac/api/tests/test_api_feec_2d.py (hunk 1 above)
Comment (yguclu) [reply psydac/api/tests/test_api_feec_2d.py:381]: Thanks for noticing that! Done in commit b2023c9

File: psydac/api/fem.py [hunk 8]
```diff
@@ -213,22 +213,33 @@ def __init__(self, expr, kernel_expr, *args, **kwargs):
             trial_ext = None
 
         if isinstance(domain, (sym_Boundary, sym_Interface)):
+
+            #...
+            # If process does not own the boundary or interface, do not assemble anything
+            import psydac
```
Comment (saidctb) [inline psydac/api/fem.py:219]: you forgot this import

File: psydac/api/fem.py [hunk 9]
```diff
@@ -459,15 +470,26 @@ def __init__(self, expr, kernel_expr, *args, **kwargs):
         else:
             ext  = domain.ext
             axis = domain.axis
-            if ext == -1:
-                start = self.space.vector_space.starts[domain.axis]
-                if start != 0 :
+
+            #...
+            # If process does not own the boundary or interface, do not assemble anything
+            import psydac
```
Comment (saidctb) [inline psydac/api/fem.py:476]: same here

File: psydac/api/fem.py (hunk 8 above)
Comment (yguclu) [reply psydac/api/fem.py:219]: It's used one line below

File: psydac/api/fem.py (hunk 9 above)
Comment (yguclu) [reply psydac/api/fem.py:476]: It's used one line below

File: psydac/feec/global_projectors.py (hunk 2 above)
Comment (yguclu) [reply psydac/feec/global_projectors.py:110]: Yes, `dim` should be read from the space, not from the quadrature rule.

**Comment by yguclu** [comment]
I forgot that we have a `DenseMatrix` class that needs to be updated. I will fix it soon.

File: psydac/feec/global_projectors.py (hunk 2 above)
Comment (yguclu) [reply psydac/feec/global_projectors.py:113]: Fixed in commit fcfed12

File: psydac/feec/global_projectors.py (hunk 3 above)
Comment (yguclu) [reply psydac/feec/global_projectors.py:280]: This was an error, I fixed it in commit 3367105

File: psydac/feec/global_projectors.py (hunk 4 above)
Comment (yguclu) [reply psydac/feec/global_projectors.py]: Yes indeed. I fixed it in commit 1198ebb

File: psydac/feec/global_projectors.py (hunk 7 above)
Comment (yguclu) [reply psydac/feec/global_projectors.py]: @campospinto I've updated the docstrings of the H1 projector in commit 6457ee9. Please let me know what you think of it!

I will then proceed with updating the docstrings of the other projectors.

Comment (campospinto) [reply psydac/feec/global_projectors.py]: Very clear now, thanks

File: psydac/feec/global_projectors.py (hunk 6 above)
Comment (yguclu) [reply psydac/feec/global_projectors.py:470]: Done for the global projectors. Functions like `evaluate_dof_0form_1d` are private to the module, and do not need a docstring at this stage.

File: psydac/feec/global_projectors.py (hunk 7 above)
Comment (yguclu) [reply psydac/feec/global_projectors.py]: I updated the docstrings of the other projectors in commit aceccae.

**Comment by yguclu** [comment]
@saidctb @campospinto Please see the updated PR description for a list of all the changes made.

**Review by saidctb — APPROVED** [review]

**Review by campospinto — APPROVED** [review]
seems good to me!
